pkg/kademlia/pool.go

```diff
@@ -0,0 +1,209 @@
+// Copyright (C) 2019 Storj Labs, Inc.
+// See LICENSE for copying information.
+
+package kademlia
+
+import (
+	"context"
+	"sync"
+
+	"github.com/zeebo/errs"
+	"go.uber.org/zap"
+	"google.golang.org/grpc"
+
+	"storj.io/storj/internal/sync2"
+	"storj.io/storj/pkg/pb"
+	"storj.io/storj/pkg/transport"
+)
+
+// Pool is a kademlia connection pool.
+type Pool struct {
+	log       *zap.Logger
+	transport transport.Client
+	size      int
+
+	limit  sync2.Semaphore
+	mu     sync.Mutex
+	closed bool
+	recent []*Conn
+}
+
+// Conn represents a cached kademlia cache.
+type Conn struct {
+	pool     *Pool
+	refcount int32 // only modify when holding pool.mu
```

Comment:
oh whoa uh, okay

Reply:
i feel like the right thing here is to have conn not have the refcount, but have Pool have a mapping of Conns to refcounts internally